File: output/StarCoder/codenet/C++/Java/s314711832.java
import java.util.*;

public class s314711832 {

  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);
    int V = sc.nextInt();
    int E = sc.nextInt();
    int r = sc.nextInt();

    int[] distances = new int[V];
    Arrays.fill(distances, Integer.MAX_VALUE);
    distances[r] = 0;

    for (int i = 0; i < E; i++) {
      int source = sc.nextInt();
      int target = sc.nextInt();
      int distance = sc.nextInt();
      distances[target] = Math.min(distances[target], distances[source] + distance);
    }

    for (int i = 0; i < V; i++) {
      if (distances[i] == Integer.MAX_VALUE) {
        System.out.println("INF");
      } else {
        System.out.println(distances[i]);
      }
    }
  }
}

File: output/StarCoder/codenet/C++/Python/s314711832.py
import sys

def main():
    V, E, r = map(int, sys.stdin.readline().split())

    edges = []

    for i in range(E):
        source, target, distance = map(int, sys.stdin.readline().split())
        edges.append((source, target, distance))

    distances = [float('inf')] * V
    distances[r] = 0

    for i in range(V):
        for source, target, weight in edges:
            if distances[source] + weight < distances[target]:
                distances[target] = distances[source] + weight

    for i in range(V):
        if distances[i] == float('inf'):
            print('INF')
        else:
            print(distances[i])


if __name__ == '__main__':
    main()